crates/rune-cli/src/main.rs

```diff
@@ -95,119 +202,138 @@ struct Args {
-    ///
-    /// debug-info[=<true/false>] - Enable or disable debug info.
+
+    #[structopt(flatten)]
+    shared: SharedArgs,
+}
+
+#[derive(Debug, Clone, StructOpt)]
+#[structopt(name = "rune", about = "The Rune Language Interpreter", version = VERSION)]
+struct Args {
+    /// Control if output is colored or not.
     ///
-    /// macros[=<true/false>] - Enable or disable macros (experimental).
+    /// Valid options are:
+    /// * `auto` - try to detect automatically.
+    /// * `ansi` - unconditionally emit ansi control codes.
+    /// * `always` - always enabled.
     ///
-    /// bytecode[=<true/false>] - Enable or disable bytecode caching (experimental).
-    #[structopt(name = "option", short = "O", number_of_values = 1)]
-    compiler_options: Vec<String>,
+    /// Anything else will disable coloring.
+    #[structopt(short = "C", long, default_value = "auto")]
+    color: String,
+
+    /// The command to execute
+    #[structopt(subcommand)] // Note that we mark a field as a subcommand
+    cmd: Command,
 }
 
-async fn try_main() -> Result<ExitCode> {
-    env_logger::init();
+impl Args {
+    /// Construct compiler options from cli arguments.
+    fn options(&self) -> Result<rune::Options, rune::ConfigurationError> {
+        let mut options = rune::Options::default();
 
-    let mut args = Args::from_args();
+        // Command-specific override defaults.
+        match &self.cmd {
+            Command::Test(_) | Command::Check(_) => {
+                options.test(true);
+                options.bytecode(false);
+            }
+            Command::Run(_) => (),
+        }
 
-    if args.dump {
-        args.dump_unit = true;
-        args.dump_stack = true;
-        args.dump_functions = true;
-        args.dump_types = true;
-        args.dump_native_functions = true;
-        args.dump_native_types = true;
+        for option in &self.shared().compiler_options {
+            options.parse_option(option)?;
+        }
+
+        Ok(options)
     }
 
-    if args.dump_unit {
-        args.dump_unit = true;
-        args.dump_instructions = true;
+    /// Access shared arguments.
+    fn shared(&self) -> &SharedArgs {
+        match &self.cmd {
+            Command::Check(args) => &args.shared,
+            Command::Test(args) => &args.shared,
+            Command::Run(args) => &args.shared,
+        }
     }
 
-    if args.dump_functions
-        || args.dump_native_functions
-        || args.dump_stack
-        || args.dump_types
-        || args.dump_instructions
-    {
-        args.dump_unit = true;
+    /// Access shared arguments mutably.
+    fn shared_mut(&mut self) -> &mut SharedArgs {
+        match &mut self.cmd {
+            Command::Check(args) => &mut args.shared,
+            Command::Test(args) => &mut args.shared,
+            Command::Run(args) => &mut args.shared,
+        }
     }
+}
 
-    let mut options = rune::Options::default();
+const SPECIAL_FILES: &[&str] = &[
```

Comment:
This is the only part I'm a little bit skeptical over right now. All the entrypoints seem like potentially surprising, and they depend on where the command is executed rather than a well-defined project root.

I'm fine with keeping them for now though (maybe consider trimming) and superseding them with a more explicit mechanism.